### PR TITLE
[SPARK-13801][SQL] DataFrame.col should return unresolved attribute

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Let's start with an example:

```
val df = ...
val df2 = df.filter(...)
df.join(df2, (df("key") + 1) === df2("key"))
```

This query won't work and returns wrong result as `df("key")` and `df2("key")` reference to a same column.

I think the biggest problem is, we give users the resolved attribute. However, resolved attribute is not real column, as logical plan's output may change. For example, we will generate new output for the right child in self-join.

We should make `Dataset.col` return unresolved attribute, and still do the resolution to make sure the given column name is resolvable, but don't return the resolved one, just get the name out and wrap it with UnresolvedAttribute.

We should also alias every single `Dataset` ever generated, with a globally unique name, so that `Dataset.col` will return an unresolved attribute with a unique qualifier, and our framework can reference the corrected column with it.

When we alias a `Dataset`, we should clean out previously generated aliases, to avoid making the plan tree too big.

Now if users run the query again, it works.

There are 3 exceptional cases that we should not alias:
1. `Dataset.as`: this is obvious, we can't override user-provided alias.
2. `Dataset.toDF()`: this just turns a strongly typed `Dataset` to generic `DataFrame`, not creating a new one, no need to alias.
3. `join`: `join` is very special, it may have same-name outputs with different qualifiers, so we can't alias it as we will replace the original qualifiers and make the same-name outputs un-distinguishable. We can't clean out previously generated aliases either, as it will erase the original qualifiers.

However, if users manually introduce ambiguity, this PR will break existing code, for example:

```
val df1 = ....  // schema: [id, name]
val df2 = ....  // schema: [id, age]
val left = df1.as("same")
val right = df2.as("same")
left.join(right, left("id") === right("id"))
```

It will fail analysis after this PR, because the 2 `id` columns have same qualifier.

This PR also breaks some corner cases that works before, but I think it's good to break it:

```
val df = ...
val df1 = df.filter(...)
df1.select(df("key"))
```

It will throw an exception now, because we can't resolve `df("key")` based on `df1`. I think this is reasonable, `df` and `df1` are different `Dataset`s, it's weird to support select a column from different `Dataset`. Another case:

```
val df = ...
df.join(df, df("key") === df("key"))
```

It doesn't make sense to support self-join in this way, users should use `df.join(df, "key")`. And this trick is very fragile, it stops working if we do some operation on the column, e.g. `df.join(df, (df("key") + 1) === df("key"))`
## How was this patch tested?

existing tests.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameJoinSuite.scala [hunk 1]
```diff
@@ -115,7 +115,7 @@ class DataFrameJoinSuite extends QueryTest with SharedSQLContext {
       Row("1", 1) :: Row("2", 1) :: Row("3", 1) :: Nil)
   }
 
-  test("[SPARK-6231] join - self join auto resolve ambiguity") {
+  ignore("[SPARK-6231] join - self join auto resolve ambiguity") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameJoinSuite.scala]: With this change, I'm not sure this "auto resolve ambiguity" feature is useful anymore.

**Comment by cloud-fan** [comment]
cc @rxin @marmbrus @yhuai

**Comment by liancheng** [comment]
The major problem I see is that now any two DataFrames must be aliased first before being joined. This breaks existing user code. However, this PR does solve a more common and subtle problem that is hard to debug, especially when the join condition is not as trivial as `df("i") === df2("i")`. Our existing strategy only issues intuitive exception message against trivial comparisons.

**Comment by gatorsmile** [comment]
I like the idea. If we do not change the current behavior, name ambiguity is just confusing. Maybe we can accept it. However, more importantly, it could generate a wrong result.  See the example

```scala
val df1 = Seq((1, 3), (2, 1)).toDF("keyCol1", "keyCol2")
val df2 = Seq((1, 4), (2, 1)).toDF("keyCol1", "keyCol3")
val df3 = df1.join(df2, df1("keyCol1") === df2("keyCol1")).select(df1("keyCol1"), $"keyCol3")
df3.join(df1, df3("keyCol3") === df1("keyCol1")).show()
```
The above query returns an empty result set.
However, the correct result should be (2,1,1,3).

**Comment by cloud-fan** [comment]
cc @rxin @marmbrus , it's ready for review :)

**Comment by marmbrus** [comment]
I did a quick pass over this, and it seems reasonable.  However, this is a pretty big change so we should make sure we aren't missing anything.  It would be very helpful if you could augment the description with some more examples, focusing on the following:
 - code that used to run, but will now throw an exception
 - cases that didn't work before but do now
 - case where we still can't resolve as the user expects (but are hopefully giving a better error message)

You might even look through JIRA/the user list to help enumerate cases.

**Comment by gatorsmile** [comment]
Let me copy the related JIRA lists I found:

- https://issues.apache.org/jira/browse/SPARK-11231
- https://issues.apache.org/jira/browse/SPARK-11576
- https://issues.apache.org/jira/browse/SPARK-10925
- https://issues.apache.org/jira/browse/SPARK-10838
- https://issues.apache.org/jira/browse/SPARK-7428

**Comment by gatorsmile** [comment]
One more related issue was just opened:

- https://issues.apache.org/jira/browse/SPARK-14040

**Comment by cloud-fan** [comment]
Hi @gatorsmile , thanks very much for listing these JIRAs! I'll go through them and augment the description later :)

**Comment by maropu** [comment]
It seems that user intentions in the example query is obvious to some extent, so I wonder if we can automatically fix this kind of name ambiguity in an `Analyzer` phase?
Throwing an exception to make users eliminate ambiguity is one of solutions though,
we slightly loss the simplicity of user codes.

**Comment by aray** [comment]
While this may help with join ambiguity. I think the more fundamental problem is that a transformed DataFrame should not be giving the same column references as the original. For an example where this is still a problem even with this patch see @wesm recent blog post http://wesmckinney.com/blog/compiling-dataframe-code/

**Comment by cloud-fan** [comment]
cc @maropu @aray , yea, we can do better for this problem, this PR is just step 1. We can alias every DataFrame so that transformed DataFrame won't give the same column references as the original. But it requires more changes and I'd like to do it a follow-ups

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
retest this please

**Comment by liancheng** [comment]
Shall we close this?

**Pull request closed without merging**